File: src/mt/edu/um/cf2/jgribx/GribRecord.java
/*
 * ============================================================================
 * JGribX
 * ============================================================================
 * Written by Andrew Spiteri <andrew.spiteri@um.edu.mt>
 * Adapted from JGRIB: http://jgrib.sourceforge.net/
 * 
 * Licensed under MIT: https://github.com/spidru/JGribX/blob/master/LICENSE
 * ============================================================================
 */
package mt.edu.um.cf2.jgribx;

import java.io.IOException;
import java.util.Calendar;
import java.util.List;
import mt.edu.um.cf2.jgribx.grib1.Grib1Record;
import mt.edu.um.cf2.jgribx.grib2.Grib2Record;

public abstract class GribRecord
{
    private GribRecordIS is;
    
    public static GribRecord readFromStream(GribInputStream in) throws IOException, NotSupportedException, NoValidGribException
    {
        GribRecord record = null;
        
        GribRecordIS is = GribRecordIS.readFromStream(in);
        if (is == null) return null;
        
        switch (is.getGribEdition())
        {
            case 1:
                record = Grib1Record.readFromStream(in, is);
                break;
            case 2:
                record = Grib2Record.readFromStream(in, is);
                break;
            default:
                throw new NoValidGribException("Unsupported GRIB edition "+is.getGribEdition());
        }
        
        GribRecordES es = new GribRecordES(in);
        if (!es.isValid)
        {
            throw new NoValidGribException("Grib End Section is invalid");
        }
        
        if (record != null)
        {
            record.is = is;
        }
        return record;
    }
    
    public GribRecordIS getIS()
    {
        return is;
    }
    
    /**
     * Returns the ID corresponding to the originating centre.
     * @return the ID corresponding to the originating centre
     */
    public abstract int getCentreId();
    public abstract Calendar getForecastTime();
    public abstract String getLevelCode();
    public abstract String getLevelDescription();
    
    /**
     * Returns the unique ID for the level code and value combination.
     * @return 
     */
    public abstract String getLevelIdentifier();
    public abstract float[] getLevelValues();
    public abstract String getParameterCode();
    public abstract String getParameterDescription();
    
    /**
     * Returns the ID corresponding to the generating process.
     * @return the ID corresponding to the generating process
     */
    public abstract int getProcessId();
    public abstract Calendar getReferenceTime();
    public abstract double getValue(double latitude, double longitude);
}

File: src/main/java/mt/edu/um/cf2/jgribx/GribRecord.kt
/*
 * ============================================================================
 * JGribX
 * ============================================================================
 * Written by Andrew Spiteri <andrew.spiteri@um.edu.mt>
 * Adapted from JGRIB: http://jgrib.sourceforge.net/
 * 
 * Licensed under MIT: https://github.com/spidru/JGribX/blob/master/LICENSE
 * ============================================================================
 */
package mt.edu.um.cf2.jgribx

import java.util.*

/**
 * A GRIB Record is a representation of part of a [GRIB message][mt.edu.um.cf2.jgribx.api.GribMessage] containing one
 * dataset. Different GRIB editions support different number of such records per
 * [GRIB message][mt.edu.um.cf2.jgribx.api.GribMessage]. Currently GRIB1 supports exactly one records per message while
 * GRIB2 supports multiple records per message due to possible repetitions of sections 2 to 7, 3 to 7, or 4 to 7.
 *
 * @param indicatorSection [Section 0: Indicator Section](https://www.nco.ncep.noaa.gov/pmb/docs/grib2/grib2_doc/grib2_sect0.shtml)
 * @author Jan Kubovy [jan@kubovy.eu]
 */
abstract class GribRecord(val indicatorSection: GribRecordIS) {
	/** Returns the ID corresponding to the originating centre. */
	abstract val centreId: Int
	abstract val forecastTime: Calendar
	abstract val levelCode: String
	abstract val levelDescription: String

	/** Returns the unique ID for the level code and value combination. */
	abstract val levelIdentifier: String
	abstract val levelValues: FloatArray
	abstract val parameterCode: String
	abstract val parameterDescription: String

	/** Returns the ID corresponding to the generating process. */
	abstract val processId: Int
	abstract val referenceTime: Calendar
	abstract fun getValue(latitude: Double, longitude: Double): Double
}
